webservice/s3_file_handler.py

```diff
@@ -0,0 +1,134 @@
+        except ClientError as e:
+            # If a client error is thrown just return status code. For example,
+            # 404 indicates that the bucket does not exist.
+            status_code = int(e.response['Error']['Code'])
+            return status_code
+
+    def list_objects_in_bucket(self, bucket):
+        """
+        :param bucket: name of bucket in account
+        :type bucket: str
+        :return: list of objects in bucket_name
+        :rtype: list
+        """
+        response = self.session.list_objects_v2(Bucket=bucket)
+        if response.has_key('Contents'):
+            L = [response['Contents'][x]['Key']
+                 for x in range(len(response['Contents']))]
+            return [str(r) for r in L]  # convert to utf-8 (is unicode)
+        else:
+            return []
+
+    def get_bucket_list(self):
+        """
+        :returns a list of all buckets in AWS account
+        :rtype: list"""
+        buckets = self.bucket.list_buckets()
+        return [bucket['Name'] for bucket in buckets['Buckets']]
+
+    def upload_object_to_bucket(self, bucket, filename, key):
+        """Uploads some object (e.g., a file) to an S3 bucket.
+        :param bucket: name of bucket in AWS account
+        :type bucket: str
+        :param filename: absolute name of the object to upload
+        :type filename: str
+        :param key: object name in bucket on S3
+        :type key: str
+        :returns: status code, and msg if not 200
+        :rtype: JSON
+        """
+
+        status_code = self.bucket_exists(bucket)
+        if status_code == 200:
+            try:
+                # If it throws no exception we assume all went well.
+                self.resource.meta.client.upload_file(
+                    filename, bucket, key)  # executes silently
+                return {'status_code': 200}
+            except ClientError as e:
+                return {'status_code': e.response['Error']['Code']}
+        else:
+            return {'status_code': status_code}  # e.g., NoSuchBucket = 404
+
+    def create_presigned_url(self, bucket_name, key_name):
+        """
+        :return result: status code and presigned URL (in case of success)
+        :rtype: dict
+        """
+        result = {
+            'status_code': '',
+            'presigned_url': ''
+            }
+        status_code = self.bucket_exists(bucket_name)
+        if status_code == 200:
+            try:
+
+                url = self.bucket.generate_presigned_url(
+                    ClientMethod='get_object',
+                    Params={
+                        'Bucket': bucket_name,
+                        'Key': key_name
+                    }
+                )
+            except ClientError as e:
+                result['status_code'] = \
+                    {'status_code': e.response['Error']['Code']}
+                return result
+
+            response = requests.get(url)
+            if response.status_code == 200:
```

Comment:
Please see my comment in [the ticket](https://ucsc-cgl.atlassian.net/browse/AZUL-152?focusedCommentId=14139&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14139) (formatting doesn't look right in the GitHub issue) about the exact format of response.

This is why I asked you in [this comment](https://ucsc-cgl.atlassian.net/browse/AZUL-152?focusedCommentId=14263&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14263) to to summarize the query parameter and the response body format, so that we could agree on it and you could implement exactly that.

Reply:
If `status_code` isn't 200, then it returns the HTTP error status code, and both, `presigned_url` and `s3_upload_status_code` have value `null`.

Reply:
A successful response should match the Fence response that I linked to in JIRA.

Reply:
My design rested on the notion to return the same JSON, no matter what status code. I changed the happy path to Fence's successful response:
`{url: <url>}
`
so no status code (200 is implied I suppose). 
But upon encountering an error don't we want to return a more detailed JSON, providing the error? If so, how much do we need to go into detail about the error? According to the Apigee article a lot, and I will try that, but hope I'm not overdoing it. I have so far included small messages in the body, but no formal error codes or status code explanations. Let's discuss the desired level of detail. Couldn't figure out much how Fence handles this so far.

Reply:
BTW, I needed to change the URL of the Boardwalk instance. Use 
`curl 'https://mkrause3.ucsc-cgp-dev.org/api/v1/repository/files/export?filters=%7B%7D&format=bdbag' -H 'accept: application/json'`
to see the response.

Reply:
The HTTP response code itself has has to indicate success or failure. A user should not have to look at the response body to determine whether there was success of failure. Maybe you're saying that and doing that, but to me it looked like, if there was an error, you would return a status code of 200, with a 4xx or a 5xx in the response body. If I'm misreading that, and the status code of the response is 4xx or 5xx, then you are good.

Regarding the error response:

1. I would not spend a lot of time dealing with it, something close to what you have now is fine, although I don't think it helps to have an empty presigned url property.
2. I think the error response format should be generic, so that it could be used for errors for any API. That way clients can write generic error handlers, and not have to customize the error handling for each different API.

Reply:
BTW, here's how Fence handles a 4xx: https://github.com/uc-cdis/fence/blob/master/fence/errors.py#L8 and 
https://github.com/uc-cdis/fence/blob/master/fence/error_handler.py#L32

They are returning an HTML response; I think it's plenty to just return a message as they have in errors.py

Reply:
I've modified all responses to client. The response does not contain any empty strings anymore. The happy path just contains the URL. In case of an error it returns the status code, the error code (if available) and a short message.
I also added a variable that sets the expiration time of the presigned URL, with a default value of 120 s to contain misuse potential.